man/man5/rlm_unbound.5
.\"     # DS - begin display
.de DS
.RS
.nf
.sp
..
.\"     # DE - end display
.de DE
.fi
.RE
.sp
..
.TH rlm_unbound 5 "8 July 2013" "" "FreeRADIUS Module"
.SH NAME
rlm_unbound \- FreeRADIUS Module
.SH DESCRIPTION
Each instance of \fIrlm_unbound\fP provides an embedded DNS client
for performing DNS lookups.  Each instance may be configured separately
to query different DNS horizons, change DNSSEC options, etc.
.PP
The module is primarily intended for use by other modules through
internal APIs, and so, instances should be initialized earlier than
those modules which use them.  Each instance does also provide some
xlat functionalities for general use and for troubleshooting.
.PP
Each instance of rlm_unbound may take the following parameters:
.IP filename
This file must exist and must point to a valid libunbound configuration file.
The default is ${raddbdir}/mods-config/unbound/default.conf.
.IP timeout
While libunbound provides an asyncronous API for internal use, using any xlat
is done syncronously from the perspective of unlang.  This value limits the
amount of time a request will wait for DNS to respond, after which the xlat
will fail.  The default is 3000 milliseconds.  This setting is independent of
any libunbound configuration values.
.IP resolvconf
Full path of a resolv.conf file to load resolver details from.  If this is
not set then libunbound will query root DNS servers.
.IP hosts
Full path of a hosts file to load.  This provides a mechanism for local
overrides to names which would otherwise not resolve or need different
results to those which a DNS resolution would provide.
.PP
An instance named, for example, "dns" will provide the following xlat
functionalities:
.IP %{dns-a:<owner>}
Performs an A lookup for the owner name, returning a stringified IPv4
address.  Only the first A record in the RRSET will be returned.
.IP %{dns-aaaa:<owner>}
Performs an AAAA lookup for the owner name, returning a stringified IPv6
address.  Only the first AAAA record in the RRSET will be returned.
.IP %{dns-ptr:<owner>}
Performs a PTR lookup for the owner.
.PP
.SH CAVEATS
Logging from rlm_unbound can be problematic, especialy if more than one
instantiation of the module is used.  This is due to the need for additional
features in the underlying libunbound which hopefully will be enhanced over
time.
.PP
There is a potential for a FreeRADIUS server using rlm_unbound to either
fail to terminate cleanly (leaving zombie processes, failing to clean up
other modules, and hanging after a SIGTERM until a SIGKILL is sent) or
to fail valgrind checks during termination when run with -m.  Likewise this
problem will rely on upstream enhancements before it can be fixed, and the
exact behavior may change in interim releases until then.
.PP
The logging behavior of rlm_unbound may vary depending on whether
FreeRADIUS is compiled with support for threads.
.PP
.SH FILES
.I /etc/raddb/modules-available/rlm_unbound
.I /etc/raddb/modules-config/unbound/
.PP
.SH "SEE ALSO"
.BR radiusd (8),
.BR radiusd.conf (5)
.BR libunbound (3)
.BR unbound.conf (5)
.SH AUTHOR
Brian S. Julin, bjulin@clarku.edu



doc/ChangeLog
FreeRADIUS 3.0.27 Tue 20 Sep 2022 12:00:00 EDT urgency=low
	Feature improvements
	* Update dictionary.alcatel.sr

	Bug fixes
	* Fix rlm_python3 build with python >= 3.10. Fixes #4441
	* The DS-Lite-Tunnel-Name data type should be 'octets'.
	* Fix rlm_expr destroying MD context, causing leaks with OpenSSL >= 3.0 #4893

FreeRADIUS 3.0.26 Tue 20 Sep 2022 12:00:00 EDT urgency=low
	Feature improvements
	* Add support for OpenSSL3
	* Add dictionary.mellanox, dictionary.netelastic,
	  dictionary.ciena, dictionary.nile
	* Update dictionary.aruba, dictionary.roaringpenguin
	* Removed haproxy support.  It's new and experimental,
	  and belongs in the v3.2 release, where it will be
	  fully supported.
	* Support PEAP and TTLS with TLS 1.3.  This has been
	  tested with wpa_supplicant and Windows 11.
	* Add %{concat:foo[*] ;}, which concatenates a set
	  of attributes, separated by a character.
	* Added sample configuration for using Google LDAP.
	  See sites-available/google-ldap-auth, and
	  mods-available/google_ldap, and mods-available/cache_auth.
	* Add Dockerfiles for Rocky8.
	* Add raduat script to the freeradius-utils package.
	* Add Debian freeradius-freetds package.
	* Add client short name to "dropping packet" message,
	* Update MS-SQL queries to avoid using column which was
	  deleted years ago.
	* Add configure-time FIPS workaround to use internal MD4/MD5
	  implementations when disabled in OpenSSL.

	Bug fixes
	* Ensure PBKDF2 always uses at least one iteration
	* Actually use the certificate in "realm_dir" hinted at by SNI.
	* Removed the use of "locate" during the "configure" phase.
	  Fixes #4318
	* Fix for showing incomplete home server list after deleting a
	  home server via radmin.
	* Call closedir() when reading certificates from a directory.
	  Found by Antonio Torres.  Fixes #4378.
	* If we read more than 16K of data in RadSec, then read it all.
	  Fixes #4388, patch from Aren Sandersen.
	* Fix information leak in compute_password_element() function
	  of EAP-PWD.  Found by Mohamed Sabt.
	* Fix crash in EAP-SIM when unknown attributes are sent.
	* Update linelog, etc. to allow the use of /dev/stdout
	* Fix crash in race condition when 500+ sockets are open.
	* Remove sample "cache_eap" module.  It does not work, and
	  offers no benefit.
	* Fix crash in RadSec with expired server certificate. #4447.
	* PEAP now correctly runs Post-Auth-Type Accept.
	* Build fixes for OS X
	* Minor updates and fixes to CI, Dockerfiles and packaging.

FreeRADIUS 3.0.25 Thu 07 Oct 2021 12:00:00 EDT urgency=medium
	Feature improvements
	* Better debug output when proxying is disabled
	* Updates to support PostgreSQL 14 (#4251)

	Bug fixes
	* Add `correct_escapes` back into default configuration
	* Fix undeclared variable with some compile options (#4246)
	* Quiet erroneous debug output
	* Fix segfault when proxying to zombie home server
	* Fix resolving values to enum strings in rlm_rest (#4167)
	* Fix printing raw values rather than enum strings in rlm_couchbase (#4167)

FreeRADIUS 3.0.24 Wed 29 Sep 2021 12:00:00 EDT urgency=medium
	Feature improvements
	* Add sanitizer options to configure script.
	* Log information needed by Wireshark to decode TLS sessions.
	* Allow more liberal SQL commands in rlm_sql_map.
	* Update dictionary.apc, dictionary.h3c
	* Add new Acct-Status-Type Subsystem-On and Subsystem-Off.
	  See dictionary.iana and
	  https://freeradius.org/rfc/acct_status_type_subsystem.html
	* Add reject_unknown_intermediate_ca.  See mods-available/eap
	* Add dynamic loading of certificates via TLS-Session-Cert-File.
	  See raddb/certs/realms/README.md
	* Add Server Name Indication (SNI) for outbound RadSec connections.
	  See raddb/sites-available/tls, and the home server tls configuration.
	* Support SNI for inbound RadSec connections.  Certificates will
	  be loaded from "realm_dir" in the "tls" section.  SNI will be
	  cached in the TLS-Server-Name-Indication attribute.
	* Preliminary support for haproxy "PROXY" protocol.
	  See sites-available/tls, "proxy_protocol" and doc/antora/modules/howto/pages/protocols/proxy/
	* Generate parse errors in more circumstances when we know that the
	  configuration is wrong.
	* Add "weeklycounter" to sample sqlcounter configuration
	* Add certificate attributes to the request list, even if the certificates
	  have expired.
	* The Simultaneous-Use code is now IPv6 aware, and can deal with
	  NAS-IPv6-Address.
	* Add dictionary.cambium
	* Add new %{lasttime:...} allowing calculation of a number of
	  hours/days/weeks/years in the past.

	Bug fixes
	* Fix crash in trustrouter module (#4115). Patch from Alejandro Perez
	* Fix crash in state handling.
	* Don't alter global options in redhat logrotate scripts.
	* EAP-FAST will print errors and continue, rather than exiting
	  when OpenSSL fails various internal sanity checks.
	* Allow admin to manually change core limits, even when core limits
	  are disabled. Patch from Antonio Torres.
	* Fix chunked rlm_rest HTTP body.  Closes #4131.
	  Patch from Nathan Ward.
	* Many fixes around the SQL ippool queries.conf and schema.
	  Patches from Jorge Periera.
	* Fix MySQL stored procedures.  PR #4170 from Terry Burton.
	* Rework connection pool management for corner cases.
	  Fixes #4161, #4162, #4163.
	* Final fix for double free in #3188.
	* Fix sqlcounter wrong memory free. PR #4192 from Jorge Pereira
	* Accept slow writes from proxies over TCP, which allows the
	  server to make more progress when it receives partial packets.
	* Add 'weeklycounter' for rlm_sqlcounter.
	* Outbound proxying over TCP / TLS is better able to deal with
	  partial TCP reads, and has fewer issues with slow networks.
	* Fix wrong data-type of Acct-Delay-Time in rlm_unix.
	* Fix EAP-FAST PAC lifetime calculation.
	* Print correct encoded packet length when debugging

FreeRADIUS 3.0.23 Thu 10 Jun 2021 12:00:00 EDT urgency=low
	Feature improvements
	* Update dictionary.aruba
	* Add "set home_server state ... down" in order to mark the

src/modules/rlm_unbound/rlm_unbound.c
/*
 *   This program is is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or (at
 *   your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 */

/**
 * $Id$
 * @file rlm_unbound.c
 * @brief DNS services via libunbound.
 *
 * @copyright 2013 The FreeRADIUS server project
 * @copyright 2013 Brian S. Julin (bjulin@clarku.edu)
 */
RCSID("$Id$")

#define LOG_PREFIX mctx->inst->name

#include <freeradius-devel/server/base.h>
#include <freeradius-devel/server/module_rlm.h>
#include <freeradius-devel/server/log.h>
#include <fcntl.h>

#include "io.h"
#include "log.h"

typedef struct {
	char const	*name;
	uint32_t	timeout;

	char const	*filename;		//!< Unbound configuration file
	char const	*resolvconf;		//!< resolv.conf file to use
	char const	*hosts;			//!< hosts file to load
} rlm_unbound_t;

typedef struct {
	unbound_io_event_base_t	*ev_b;		//!< Unbound event base
	rlm_unbound_t		*inst;		//!< Instance data
	unbound_log_t		*u_log;		//!< Unbound log structure
} rlm_unbound_thread_t;

typedef struct {
	int			async_id;	//!< Id of async query
	request_t		*request;	//!< Current request being processed
	rlm_unbound_thread_t	*t;		//!< Thread running this request
	int			done;		//!< Indicator that the callback has been called
						///< Negative values indicate errors.
	bool			timedout;	//!< Request timedout.
	fr_type_t		return_type;	//!< Data type to parse results into
	bool			has_priority;	//!< Does the returned data start with a priority field
	uint16_t		count;		//!< Number of results to return
	FR_DLIST_HEAD(fr_value_box_list)	list;		//!< Where to put the parsed results
	TALLOC_CTX		*out_ctx;	//!< CTX to allocate parsed results in
	fr_event_timer_t const	*ev;		//!< Event for timeout
} unbound_request_t;

/*
 *	A mapping of configuration file names to internal variables.
 */
static const CONF_PARSER module_config[] = {
	{ FR_CONF_OFFSET("filename", FR_TYPE_FILE_INPUT, rlm_unbound_t, filename), .dflt = "${modconfdir}/unbound/default.conf" },
	{ FR_CONF_OFFSET("timeout", FR_TYPE_UINT32, rlm_unbound_t, timeout), .dflt = "3000" },
	{ FR_CONF_OFFSET("resolvconf", FR_TYPE_FILE_INPUT, rlm_unbound_t, resolvconf) },
	{ FR_CONF_OFFSET("hosts", FR_TYPE_FILE_INPUT, rlm_unbound_t, hosts) },
	CONF_PARSER_TERMINATOR
};

static int _unbound_request_free(unbound_request_t *ur)
{
	/*
	 *	Cancel an outstanding async unbound call if the request is being freed
	 */
	if ((ur->async_id != 0) && (ur->done == 0)) ub_cancel(ur->t->ev_b->ub, ur->async_id);

	return 0;
}

/**	Callback called by unbound when resolution started with ub_resolve_event() completes
 *
 * @param mydata	the request tracking structure set up before ub_resolve_event() was called
 * @param rcode		should be the rcode from the reply packet, but appears not to be
 * @param packet	wire format reply packet
 * @param packet_len	length of wire format packet
 * @param sec		DNSSEC status code
 * @param why_bogus	String describing DNSSEC issue if sec = 1
 * @param rate_limited	Was the request rate limited due to unbound workload
 */
static void xlat_unbound_callback(void *mydata, int rcode, void *packet, int packet_len, int sec,
				  char *why_bogus, UNUSED int rate_limited)
{
	unbound_request_t	*ur = talloc_get_type_abort(mydata, unbound_request_t);
	request_t		*request = ur->request;
	fr_dbuff_t		dbuff;
	uint16_t		qdcount = 0, ancount = 0, i, rdlength = 0;
	uint8_t			pktrcode = 0, skip = 0;
	ssize_t			used;
	fr_value_box_t		*vb;

	/*
	 *	Request has completed remove timeout event and set
	 *	async_id to 0 so ub_cancel() is not called when ur is freed
	 */
	if (ur->ev) (void)fr_event_timer_delete(&ur->ev);
	ur->async_id = 0;

	/*
	 *	Bogus responses have the "sec" flag set to 1
	 */
	if (sec == 1) {
		RERROR("%s", why_bogus);
		ur->done = -16;
		goto resume;
	}

	RHEXDUMP4((uint8_t const *)packet, packet_len, "Unbound callback called with packet [length %d]", packet_len);

	fr_dbuff_init(&dbuff, (uint8_t const *)packet, (size_t)packet_len);

	/*	Skip initial header entries */
	fr_dbuff_advance(&dbuff, 3);

	/*
	 *	Extract rcode - it doesn't appear to be passed in as a
	 *	parameter, contrary to the documentation...
	 */
	fr_dbuff_out(&pktrcode, &dbuff);
	rcode = pktrcode & 0x0f;
	if (rcode != 0) {
		ur->done = 0 - rcode;
		REDEBUG("DNS rcode is %d", rcode);
		goto resume;
	}

	fr_dbuff_out(&qdcount, &dbuff);
	if (qdcount > 1) {
		RERROR("DNS results packet with multiple questions");
		ur->done = -32;
		goto resume;
	}

	/*	How many answer records do we have? */
	fr_dbuff_out(&ancount, &dbuff);
	RDEBUG4("Unbound returned %d answers", ancount);

	/*	Skip remaining header entries */
	fr_dbuff_advance(&dbuff, 4);

	/*	Skip the QNAME */
	fr_dbuff_out(&skip, &dbuff);
	while (skip > 0) {
		if (skip > 63) {
			/*
			 *	This is a pointer to somewhere else in the the packet
			 *	Pointers use two octets
			 *	Just move past the pointer to the next label in the question
			 */
			fr_dbuff_advance(&dbuff, 1);
		} else {
			if (fr_dbuff_remaining(&dbuff) < skip) break;
			fr_dbuff_advance(&dbuff, skip);
		}
		fr_dbuff_out(&skip, &dbuff);
	}

	/*	Skip QTYPE and QCLASS */
	fr_dbuff_advance(&dbuff, 4);

	/*	We only want a limited number of replies */
	if (ancount > ur->count) ancount = ur->count;

	fr_value_box_list_init(&ur->list);

	/*	Read the answer RRs */
	for (i = 0; i < ancount; i++) {
		fr_dbuff_out(&skip, &dbuff);
		if (skip > 63) fr_dbuff_advance(&dbuff, 1);

		/*	Skip TYPE, CLASS and TTL */
		fr_dbuff_advance(&dbuff, 8);

		fr_dbuff_out(&rdlength, &dbuff);
		RDEBUG4("RDLENGTH is %d", rdlength);

		MEM(vb = fr_value_box_alloc_null(ur->out_ctx));
		switch (ur->return_type) {
		case FR_TYPE_IPV4_ADDR:
		case FR_TYPE_IPV6_ADDR:
		case FR_TYPE_OCTETS:
			if (fr_value_box_from_network(ur->out_ctx, vb, ur->return_type, NULL,
						      &dbuff, rdlength, true) < 0) {
			error:
				talloc_free(vb);
				fr_value_box_list_talloc_free(&ur->list);
				ur->done = -32;
				goto resume;
			}
			break;

		case FR_TYPE_STRING:
			if (ur->has_priority) {
				/*
				 *	This record type has a priority before the label
				 *	add the priority first as a separate box
				 */
				fr_value_box_t	*priority_vb;
				if (rdlength < 3) {
					REDEBUG("%s - Invalid data returned", ur->t->inst->name);
					goto error;
				}
				MEM(priority_vb = fr_value_box_alloc_null(ur->out_ctx));
				if (fr_value_box_from_network(ur->out_ctx, priority_vb, FR_TYPE_UINT16, NULL,
							      &dbuff, 2, true) < 0) {
					talloc_free(priority_vb);
					goto error;
				}
				fr_value_box_list_insert_tail(&ur->list, priority_vb);
			}

			/*	String types require decoding of dns format labels */
			used = fr_dns_label_to_value_box(ur->out_ctx, vb, (uint8_t const *)packet, packet_len,
							 (uint8_t const *)fr_dbuff_current(&dbuff), true, NULL);
			if (used < 0) goto error;
			fr_dbuff_advance(&dbuff, (size_t)used);
			break;

		default:
			RERROR("No meaningful output type set");
			goto error;
		}

		fr_value_box_list_insert_tail(&ur->list, vb);

	}

	ur->done = 1;

resume:
	unlang_interpret_mark_runnable(ur->request);
}

/**	Callback from our timeout event to cancel a request
 *
 */
static void xlat_unbound_timeout(UNUSED fr_event_list_t *el, UNUSED fr_time_t now, void *uctx)
{
	unbound_request_t	*ur = talloc_get_type_abort(uctx, unbound_request_t);
	request_t		*request = ur->request;

	REDEBUG("Timeout waiting for DNS resolution");
	unlang_interpret_mark_runnable(request);

	ur->timedout = true;
}

/*
 *	Xlat signal callback if an unbound request needs cancelling
 */
static void xlat_unbound_signal(xlat_ctx_t const *xctx, request_t *request, fr_state_signal_t action)
{
	unbound_request_t	*ur = talloc_get_type_abort(xctx->rctx, unbound_request_t);

	if (action != FR_SIGNAL_CANCEL) return;

	if (ur->ev) (void)fr_event_timer_delete(&ur->ev);

	RDEBUG2("Forcefully cancelling pending unbound request");
}

/*
 *	Xlat resume callback after unbound has either returned or timed out
 *	Move the parsed results to the xlat output cursor
 */
static xlat_action_t xlat_unbound_resume(UNUSED TALLOC_CTX *ctx, fr_dcursor_t *out,
					 xlat_ctx_t const *xctx,
					 request_t *request, UNUSED FR_DLIST_HEAD(fr_value_box_list) *in)
{
	fr_value_box_t		*vb;
	unbound_request_t	*ur = talloc_get_type_abort(xctx->rctx, unbound_request_t);

	/*
	 *	Request timed out
	 */
	if (ur->timedout) return XLAT_ACTION_FAIL;

#define RCODEERROR(_code, _message) case _code: \
	REDEBUG(_message, ur->t->inst->name); \
	goto error

	/*	Check for unbound errors */
	switch (ur->done) {
	case 1:
		break;

	default:
		REDEBUG("%s - Unknown DNS error", ur->t->inst->name);
	error:
		talloc_free(ur);
		return XLAT_ACTION_FAIL;

	RCODEERROR(0, "%s - No result");
	RCODEERROR(-1, "%s - Query format error");
	RCODEERROR(-2, "%s - DNS server failure");
	RCODEERROR(-3, "%s - Nonexistent domain name");
	RCODEERROR(-4, "%s - DNS server does not support query type");
	RCODEERROR(-5, "%s - DNS server refused query");
	RCODEERROR(-16, "%s - Bogus DNS response");
	RCODEERROR(-32, "%s - Error parsing DNS response");
	}

	/*
	 *	Move parsed results into xlat cursor
	 */
	while ((vb = fr_value_box_list_pop_head(&ur->list))) {
		fr_dcursor_append(out, vb);
	}

	talloc_free(ur);
	return XLAT_ACTION_DONE;
}


static xlat_arg_parser_t const xlat_unbound_args[] = {
	{ .required = true, .concat = true, .type = FR_TYPE_STRING },
	{ .required = true, .concat = true, .type = FR_TYPE_STRING },
	{ .single = true, .type = FR_TYPE_UINT16 },
	XLAT_ARG_PARSER_TERMINATOR
};

/** Perform a DNS lookup using libunbound
 *
 * @ingroup xlat_functions
 */
static xlat_action_t xlat_unbound(TALLOC_CTX *ctx, fr_dcursor_t *out,
				  xlat_ctx_t const *xctx,
				  request_t *request, FR_DLIST_HEAD(fr_value_box_list) *in)
{
	rlm_unbound_t const		*inst = talloc_get_type_abort_const(xctx->mctx->inst->data, rlm_unbound_t);
	rlm_unbound_thread_t		*t = talloc_get_type_abort(xctx->mctx->thread, rlm_unbound_thread_t);
	fr_value_box_t			*host_vb = fr_value_box_list_head(in);
	fr_value_box_t			*query_vb = fr_value_box_list_next(in, host_vb);
	fr_value_box_t			*count_vb = fr_value_box_list_next(in, query_vb);
	unbound_request_t		*ur;

	if (host_vb->vb_length == 0) {
		REDEBUG("Can't resolve zero length host");
		return XLAT_ACTION_FAIL;
	}

	MEM(ur = talloc_zero(unlang_interpret_frame_talloc_ctx(request), unbound_request_t));
	talloc_set_destructor(ur, _unbound_request_free);

	/*
	 *	Set the maximum number of records we want to return
	 */
	if ((count_vb) && (count_vb->type == FR_TYPE_UINT16) && (count_vb->vb_uint16 > 0)) {
		ur->count = count_vb->vb_uint16;
	} else {
		ur->count = UINT16_MAX;
	}

	ur->request = request;
	ur->t = t;
	ur->out_ctx = ctx;

#define UB_QUERY(_record, _rrvalue, _return, _hasprio) \
	if (strcmp(query_vb->vb_strvalue, _record) == 0) { \
		ur->return_type = _return; \
		ur->has_priority = _hasprio; \
		ub_resolve_event(t->ev_b->ub, host_vb->vb_strvalue, _rrvalue, 1, ur, \
				xlat_unbound_callback, &ur->async_id); \
	}

	/* coverity[dereference] */
	UB_QUERY("A", 1, FR_TYPE_IPV4_ADDR, false)
	else UB_QUERY("AAAA", 28, FR_TYPE_IPV6_ADDR, false)
	else UB_QUERY("PTR", 12, FR_TYPE_STRING, false)
	else UB_QUERY("MX", 15, FR_TYPE_STRING, true)
	else UB_QUERY("SRV", 33, FR_TYPE_STRING, true)
	else UB_QUERY("TXT", 16, FR_TYPE_STRING, false)
	else UB_QUERY("CERT", 37, FR_TYPE_OCTETS, false)
	else {
		REDEBUG("Invalid / unsupported DNS query type");
		return XLAT_ACTION_FAIL;
	}

	/*
	 *	unbound returned before we yielded - run the callback
	 *	This is when serving results from local data
	 */
	if (ur->async_id == 0) {
		xlat_ctx_t our_xctx = *xctx;

		our_xctx.rctx = ur;	/* Make the rctx available to the resume function */

		return xlat_unbound_resume(ctx, out, &our_xctx, request, in);
	}

	if (fr_event_timer_in(ur, ur->t->ev_b->el, &ur->ev, fr_time_delta_from_msec(inst->timeout),
			      xlat_unbound_timeout, ur) < 0) {
		REDEBUG("Unable to attach unbound timeout_envent");
		ub_cancel(t->ev_b->ub, ur->async_id);
		return XLAT_ACTION_FAIL;
	}

	return unlang_xlat_yield(request, xlat_unbound_resume, xlat_unbound_signal, ur);
}

static int mod_thread_instantiate(module_thread_inst_ctx_t const *mctx)
{
	rlm_unbound_t		*inst = talloc_get_type_abort(mctx->inst->data, rlm_unbound_t);
	rlm_unbound_thread_t	*t = talloc_get_type_abort(mctx->thread, rlm_unbound_thread_t);
	int			res;

	t->inst = inst;
	if (unbound_io_init(t, &t->ev_b, mctx->el) < 0) {
		PERROR("Unable to create unbound event base");
		return -1;
	}

	/*
	 *	Ensure unbound uses threads
	 */
	res = ub_ctx_async(t->ev_b->ub, 1);
	if (res) {
	error:
		PERROR("%s", ub_strerror(res));
		return -1;
	}

	/*
	 *	Load settings from the unbound config file
	 */
	res = ub_ctx_config(t->ev_b->ub, UNCONST(char *, inst->filename));
	if (res) goto error;

	if (unbound_log_init(t, &t->u_log, t->ev_b->ub) < 0) {
		PERROR("Failed to initialise unbound log");
		return -1;
	}

	/*
	 *	Load resolv.conf if specified
	 */
	if (inst->resolvconf) ub_ctx_resolvconf(t->ev_b->ub, inst->resolvconf);

	/*
	 *	Load hosts file if specified
	 */
	if (inst->hosts) ub_ctx_hosts(t->ev_b->ub, inst->hosts);

	/*
	 *	The unbound context needs to be "finalised" to fix its settings.
	 *	The API does not expose a method to do this, rather it happens on first
	 *	use.  A quick workround is to delete data which won't be present
	 */
	ub_ctx_data_remove(t->ev_b->ub, "notar33lsite.foo123.nottld A 127.0.0.1");

	return 0;
}

static int mod_thread_detach(module_thread_inst_ctx_t const *mctx)
{
	rlm_unbound_thread_t	*t = talloc_get_type_abort(mctx->thread, rlm_unbound_thread_t);

	talloc_free(t->u_log);
	talloc_free(t->ev_b);

	return 0;
}

static int mod_bootstrap(module_inst_ctx_t const *mctx)
{
	rlm_unbound_t	*inst = talloc_get_type_abort(mctx->inst->data, rlm_unbound_t);
	xlat_t		*xlat;

	inst->name = mctx->inst->name;

	if (inst->timeout > 10000) {
		cf_log_err(mctx->inst->conf, "timeout must be 0 to 10000");
		return -1;
	}

	if(!(xlat = xlat_register_module(NULL, mctx, mctx->inst->name, xlat_unbound, FR_TYPE_VOID, 0))) return -1;
	xlat_func_args(xlat, xlat_unbound_args);

	return 0;
}

extern module_rlm_t rlm_unbound;
module_rlm_t rlm_unbound = {
	.common = {
		.magic			= MODULE_MAGIC_INIT,
		.name			= "unbound",
		.type			= MODULE_TYPE_THREAD_SAFE,
		.inst_size		= sizeof(rlm_unbound_t),
		.config			= module_config,
		.bootstrap		= mod_bootstrap,

		.thread_inst_size	= sizeof(rlm_unbound_thread_t),
		.thread_inst_type	= "rlm_unbound_thread_t",
		.thread_instantiate	= mod_thread_instantiate,
		.thread_detach		= mod_thread_detach
	}
};
